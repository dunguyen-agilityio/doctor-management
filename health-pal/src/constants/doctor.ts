export const DoctorQueryKey = {
  hospital: 'populate[clinic][populate]',
  doctor: 'populate[users_permissions_user][populate]',
  specialty: 'populate[specialty][populate]',
  query: 'filters[users_permissions_user][name][$containsi]',
  filterSpecialty: 'filters[specialty][name][$eqi]',
}

export const DoctorMessages = {
  DOCTOR_ERROR: 'Failed to get doctor',
  DOCTOR_LIST_ERROR: 'Failed to get doctors list',
  DOCTOR_LIST_EMPTY: 'No doctors found',
  DOCTOR_LIST_EMPTY_SEARCH: 'No doctors found with this search term',
  DOCTOR_LIST_EMPTY_SPECIALTY: 'No doctors found in this specialty',
}
