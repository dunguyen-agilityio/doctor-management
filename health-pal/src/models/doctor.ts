import { Clinic } from './clinic'
import { Review } from './review'
import { Specialty } from './specialty'
import { User } from './user'

export class Doctor extends User {
  specialty: Specialty
  clinic!: Clinic
  reviews?: Review[]
  rating: number
  reviewCounter: number

  constructor({ specialty, reviews, rating, reviewCounter, ...doctor }: Doctor) {
    super(doctor)
    this.specialty = specialty ?? []
    this.reviews = reviews ?? []
    this.rating = rating ?? 0
    this.reviewCounter = reviewCounter ?? 0
    Object.assign(this, doctor)
  }
}

export type TDoctorCard = Pick<TDoctorData, 'name' | 'documentId' | 'rating' | 'id'> & {
  avatar: string
  specialty: string
  address: string
  reviewCounter: number
}

export type TDoctorData = Doctor & {
  users_permissions_user: User
  favoriteId?: string
}
