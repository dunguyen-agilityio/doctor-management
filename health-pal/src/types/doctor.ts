export type TDoctorBanner = {
  image?: string
  title: string
  description: string
}
