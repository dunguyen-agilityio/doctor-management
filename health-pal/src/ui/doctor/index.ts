export { default as DoctorContainer } from './doctor-container'
export { default as DoctorDetailHeader } from './doctor-header'
export { default as MultipleSelectSpecialty } from './multiple-select-specialty'
