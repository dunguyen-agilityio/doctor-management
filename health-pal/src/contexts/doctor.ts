import { createContext } from 'react'

import { DoctorData } from '@app/models/doctor'

export const DoctorContext = createContext<DoctorData[]>([])
