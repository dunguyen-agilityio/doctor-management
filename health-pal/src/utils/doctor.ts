import { DoctorData, TDoctorCard } from '@app/models/doctor'

export const formatDoctor = ({
  clinic,
  specialty,
  users_permissions_user,
  ...rest
}: DoctorData): TDoctorCard => {
  const { avatar, ...user } = users_permissions_user

  return {
    ...user,
    ...rest,
    address: clinic.address,
    avatar: avatar?.url ?? '',
    specialty: specialty.name,
  }
}
