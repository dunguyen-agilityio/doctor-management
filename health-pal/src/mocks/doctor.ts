import { TDoctorData } from '@/models/doctor'

import { TDoctorBanner } from '@/types/doctor'

import { MOCK_HOSPITALS } from './hospital'
import { MOCK_REVIEWS } from './review'
import { MOCK_SPECIALTY } from './specialty'
import { MOCK_USER } from './user'

export const MOCK_DOCTORS: TDoctorData[] = [
  {
    ...MOCK_USER,
    documentId: 'doc_001',
    id: 1,
    name: 'Dr. John Smith',
    reviewCounter: 5,
    rating: 4.8,
    clinic: MOCK_HOSPITALS[0],
    dateOfBirth: new Date('1980-01-01'),
    email: 'john.smith@example.com',
    specialty: MOCK_SPECIALTY,
    avatar: {
      id: 1,
      url: 'https://picsum.photos/id/91/3504/2336',
    },
    reviews: MOCK_REVIEWS,
    favoriteId: 'fav_001',
    users_permissions_user: MOCK_USER,
  },
]

export const DOCTOR_BANNER: TDoctorBanner[] = [...new Array(6).keys()].map(() => ({
  title: `Looking for\nSpecialist Doctors?`,
  description: `Schedule an appointment with\nour top doctors.`,
}))
