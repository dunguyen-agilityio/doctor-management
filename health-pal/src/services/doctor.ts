import { APP_TOKEN } from '@app/constants'
import { DOCTOR_QUERY_KEY } from '@app/constants/doctor'

import { StrapiPagination, StrapiParams } from '@app/types/strapi'

import { MOCK_REVIEWS } from '@app/mocks/reivew'
import { TDoctorData } from '@app/models/doctor'
import { buildStrapiQuery } from '@app/utils/strapi'

import { apiClient } from './http-client'

export const getDoctors = async ({ filters = [], ...params }: StrapiParams) => {
  const searchParams = buildStrapiQuery({
    ...params,
    filters: [
      { key: DOCTOR_QUERY_KEY.avatar, query: '*' },
      { key: DOCTOR_QUERY_KEY.hospital, query: '*' },
      { key: DOCTOR_QUERY_KEY.specialty, query: 'name' },
      ...filters,
    ],
  })

  const response = await apiClient.get<StrapiPagination<TDoctorData>>(`doctors?${searchParams}`, {
    jwt: APP_TOKEN,
  })
  return response
}

export const getDoctor = async (id: string) => {
  const searchParams = buildStrapiQuery({
    filters: [
      { key: DOCTOR_QUERY_KEY.avatar, query: '*' },
      { key: DOCTOR_QUERY_KEY.address, query: 'address' },
      { key: DOCTOR_QUERY_KEY.specialty, query: 'name' },
    ],
  })

  const response = await apiClient.get<{ data: TDoctorData }>(`doctors/${id}?${searchParams}`, {
    jwt: APP_TOKEN,
  })

  return {
    ...response.data,
    summary: { experience: 2, patients: 20000, rating: 4.3, reviews: 1200 },
    reviews: MOCK_REVIEWS,
  }
}
