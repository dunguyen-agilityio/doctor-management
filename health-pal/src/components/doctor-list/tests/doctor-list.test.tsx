import { render } from '@utils-test'

import { Stack } from 'tamagui'

import { MOCK_DOCTORS } from '@app/mocks/doctor'
import { TDoctorData } from '@app/models/doctor'

import DoctorList from '../'

describe('DoctorList', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot with empty data', () => {
    const renderItem = ({ item }: { item: TDoctorData }) => (
      <Stack testID={`doctor-${item.documentId}`}>{item.name}</Stack>
    )
    const { toJSON } = render(<DoctorList data={[]} renderItem={renderItem} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('matches snapshot with custom renderItem', () => {
    const renderItem = ({ item }: { item: TDoctorData }) => (
      <Stack testID={`doctor-${item.documentId}`}>{item.name}</Stack>
    )
    const { toJSON } = render(<DoctorList data={MOCK_DOCTORS} renderItem={renderItem} />)
    expect(toJSON()).toMatchSnapshot()
  })
})
