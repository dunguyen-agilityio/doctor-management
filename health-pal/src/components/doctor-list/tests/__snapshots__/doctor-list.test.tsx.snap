// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoctorList matches snapshot with custom renderItem 1`] = `
[
  <View
    aria-label="Notifications (F8)"
    pointerEvents="box-none"
    role="region"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
    tabIndex={-1}
  >
    <View
      focusable={false}
      pointerEvents="box-none"
      style={
        {
          "flexDirection": "column",
          "maxWidth": "100%",
          "position": "absolute",
        }
      }
    />
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    collapsable={false}
    contentContainerStyle={
      {
        "paddingBottom": 8,
        "paddingHorizontal": 24,
      }
    }
    data={
      [
        {
          "avatar": {
            "id": 1,
            "url": "https://picsum.photos/id/91/3504/2336",
          },
          "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
          "clinic": {
            "address": "123 Main St, Springfield, USA",
            "description": "A premier healthcare facility offering top-tier services in Springfield.",
            "doctors": [],
            "documentId": "123e4567-e89b-12d3-a456-426614174000",
            "id": 1,
            "image": {
              "alt": "Sunrise Hospital",
              "id": 1,
              "url": "https://picsum.photos/200",
            },
            "name": "Sunrise Hospital",
            "rating": 4.2,
            "reivewCouter": 300,
            "reviews": [],
            "type": "Clinic",
          },
          "dateOfBirth": 1980-01-01T00:00:00.000Z,
          "documentId": "doc_001",
          "email": "john.smith@example.com",
          "favoriteId": "fav_001",
          "gender": true,
          "id": 1,
          "name": "Dr. John Smith",
          "nickname": "jdoe",
          "rating": 4.8,
          "reviewCounter": 5,
          "reviews": [
            {
              "comment": "Dr. Patel is a true professional who genuinely cares about his patients. I highly recommend Dr. Patel to",
              "documentId": "doc-1",
              "id": 1,
              "rating": 5,
            },
            {
              "comment": "Dr. Patel is a true professional who genuinely cares about his patients. I highly recommend Dr. Patel to",
              "documentId": "doc-2",
              "id": 2,
              "image": {
                "testUri": "../../../assets/images/user01.png",
              },
              "name": "Emily Anderson",
              "rating": 5,
            },
            {
              "comment": "Dr. Patel is a true professional who genuinely cares about his patients. I highly recommend Dr. Patel to",
              "documentId": "doc-3",
              "id": 3,
              "image": {
                "testUri": "../../../assets/images/user01.png",
              },
              "name": "Emily Anderson",
              "rating": 5,
            },
            {
              "comment": "Dr. Patel is a true professional who genuinely cares about his patients. I highly recommend Dr. Patel to",
              "documentId": "doc-4",
              "id": 4,
              "image": {
                "testUri": "../../../assets/images/user01.png",
              },
              "name": "Emily Anderson",
              "rating": 5,
            },
          ],
          "specialty": {
            "description": "Specialized in heart-related conditions and treatments.",
            "doctors": [],
            "documentId": "123e4567-e89b-12d3-a456-426614174000",
            "id": 1,
            "name": "Cardiology",
          },
          "username": "jdoe",
          "users_permissions_user": {
            "avatar": {
              "id": 1,
              "url": "https://example.com/avatar.jpg",
            },
            "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
            "dateOfBirth": 1990-01-01T00:00:00.000Z,
            "documentId": "123e4567-e89b-12d3-a456-426614174000",
            "email": "jdoe@example.com",
            "gender": true,
            "id": 1,
            "name": "John Doe",
            "nickname": "jdoe",
            "username": "jdoe",
          },
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    handlerTag={2}
    handlerType="NativeViewGestureHandler"
    initialNumToRender={20}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={0.0001}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
    windowSize={5}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          testID="doctor-doc_001"
        >
          Dr. John Smith
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;

exports[`DoctorList matches snapshot with empty data 1`] = `
[
  <View
    aria-label="Notifications (F8)"
    pointerEvents="box-none"
    role="region"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
    tabIndex={-1}
  >
    <View
      focusable={false}
      pointerEvents="box-none"
      style={
        {
          "flexDirection": "column",
          "maxWidth": "100%",
          "position": "absolute",
        }
      }
    />
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    collapsable={false}
    contentContainerStyle={
      {
        "paddingBottom": 8,
        "paddingHorizontal": 24,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    initialNumToRender={20}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={0.0001}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
    windowSize={5}
  >
    <View />
  </RCTScrollView>,
]
`;
